### Week 9
### DSC 550
### Abed Tabbalat

In [17]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import plot_confusion_matrix
import sklearn.metrics as metrics
from sklearn.neighbors import KNeighborsClassifier
from sklearn import datasets
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.model_selection import GridSearchCV
from sklearn import preprocessing
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

In [4]:
# Import the dataset and ensure that it loaded properly.
df = pd.read_csv('Loan_Train.csv')
df

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849                0.0         NaN             360.0   
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History Property_Area Loan_Status  
0               1.0         Urban           Y  
1               1.0         Rural           N  
2               1.0         Urban           Y  
3               1.0         Urban           Y  
4               1.0         Urban           Y  
..              ...           ...         ...  
609             1.0         Rural           Y  
610             1.0         Rural           Y  
611             1.0         Urban           Y  
612             1.0         Urban           Y  
613             0.0     Semiurban           N  

[614 rows x 13 columns]

In [5]:
# Drop the column “Load_ID.”
df = df.drop(['Loan_ID'], axis=1)
df

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
0      Male      No          0      Graduate            No             5849   
1      Male     Yes          1      Graduate            No             4583   
2      Male     Yes          0      Graduate           Yes             3000   
3      Male     Yes          0  Not Graduate            No             2583   
4      Male      No          0      Graduate            No             6000   
..      ...     ...        ...           ...           ...              ...   
609  Female      No          0      Graduate            No             2900   
610    Male     Yes         3+      Graduate            No             4106   
611    Male     Yes          1      Graduate            No             8072   
612    Male     Yes          2      Graduate            No             7583   
613  Female      No          0      Graduate           Yes             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                  0.0         NaN             360.0             1.0   
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

    Property_Area Loan_Status  
0           Urban           Y  
1           Rural           N  
2           Urban           Y  
3           Urban           Y  
4           Urban           Y  
..            ...         ...  
609         Rural           Y  
610         Rural           Y  
611         Urban           Y  
612         Urban           Y  
613     Semiurban           N  

[614 rows x 12 columns]

In [6]:
# Drop any rows with missing data.
df = df.dropna()
df

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
1      Male     Yes          1      Graduate            No             4583   
2      Male     Yes          0      Graduate           Yes             3000   
3      Male     Yes          0  Not Graduate            No             2583   
4      Male      No          0      Graduate            No             6000   
5      Male     Yes          2      Graduate           Yes             5417   
..      ...     ...        ...           ...           ...              ...   
609  Female      No          0      Graduate            No             2900   
610    Male     Yes         3+      Graduate            No             4106   
611    Male     Yes          1      Graduate            No             8072   
612    Male     Yes          2      Graduate            No             7583   
613  Female      No          0      Graduate           Yes             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
5               4196.0       267.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

    Property_Area Loan_Status  
1           Rural           N  
2           Urban           Y  
3           Urban           Y  
4           Urban           Y  
5           Urban           Y  
..            ...         ...  
609         Rural           Y  
610         Rural           Y  
611         Urban           Y  
612         Urban           Y  
613     Semiurban           N  

[480 rows x 12 columns]

In [7]:
## Checking dataframe shape:
df.shape

(480, 12)

In [9]:
# Convert the categorical features into dummy variables.
## Since Loan_Status is the target, will convert the Y/N to 1 and 0 before applying the dummy variable function
df['Loan_Status_Num'] = df['Loan_Status'].replace(to_replace=['Y', 'N'], value=[1,0])
df

<ipython-input-9-0acc725d5494>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Loan_Status_Num'] = df['Loan_Status'].replace(to_replace=['Y', 'N'], value=[1,0])


Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
1      Male     Yes          1      Graduate            No             4583   
2      Male     Yes          0      Graduate           Yes             3000   
3      Male     Yes          0  Not Graduate            No             2583   
4      Male      No          0      Graduate            No             6000   
5      Male     Yes          2      Graduate           Yes             5417   
..      ...     ...        ...           ...           ...              ...   
609  Female      No          0      Graduate            No             2900   
610    Male     Yes         3+      Graduate            No             4106   
611    Male     Yes          1      Graduate            No             8072   
612    Male     Yes          2      Graduate            No             7583   
613  Female      No          0      Graduate           Yes             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
5               4196.0       267.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

    Property_Area Loan_Status  Loan_Status_Num  
1           Rural           N                0  
2           Urban           Y                1  
3           Urban           Y                1  
4           Urban           Y                1  
5           Urban           Y                1  
..            ...         ...              ...  
609         Rural           Y                1  
610         Rural           Y                1  
611         Urban           Y                1  
612         Urban           Y                1  
613     Semiurban           N                0  

[480 rows x 13 columns]

In [10]:
## Applying dummy variables
df = pd.get_dummies(df)
df

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
5               5417             4196.0       267.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History  Loan_Status_Num  Gender_Female  Gender_Male  Married_No  \
1               1.0                0              0            1           0   
2               1.0                1              0            1           0   
3               1.0                1              0            1           0   
4               1.0                1              0            1           1   
5               1.0                1              0            1           0   
..              ...              ...            ...          ...         ...   
609             1.0                1              1            0           1   
610             1.0                1              0            1           0   
611             1.0                1              0            1           0   
612             1.0                1              0            1           0   
613             0.0                0              1            0           1   

     Married_Yes  ...  Dependents_3+  Education_Graduate  \
1              1  ...              0                   1   
2              1  ...              0                   1   
3              1  ...              0                   0   
4              0  ...              0                   1   
5              1  ...              0                   1   
..           ...  ...            ...                 ...   
609            0  ...              0                   1   
610            1  ...              1                   1   
611            1  ...              0                   1   
612            1  ...              0                   1   
613            0  ...              0                   1   

     Education_Not Graduate  Self_Employed_No  Self_Employed_Yes  \
1                         0                 1                  0   
2                         0                 0                  1   
3                         1                 1                  0   
4                         0                 1                  0   
5                         0                 0                  1   
..                      ...               ...                ...   
609                       0                 1                  0   
610                       0                 1                  0   
611                       0                 1                  0   
612                       0                 1                  0   
613                       0                 0                  1   

     Property_Area_Rural  Property_Area_Semiurban  Property_Area_Urban  \
1                      1                        0                    0   
2                      0                        0                    1   
3                      0                        0                    1   
4                      0                        0                    1   
5                      0                        0                    1   
..                   ...                      ...                  ...   
609                    1                        0                    0   
610                    1             

In [12]:
## Column name check
df.columns

Index(['ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Loan_Status_Num',
       'Gender_Female', 'Gender_Male', 'Married_No', 'Married_Yes',
       'Dependents_0', 'Dependents_1', 'Dependents_2', 'Dependents_3+',
       'Education_Graduate', 'Education_Not Graduate', 'Self_Employed_No',
       'Self_Employed_Yes', 'Property_Area_Rural', 'Property_Area_Semiurban',
       'Property_Area_Urban', 'Loan_Status_N', 'Loan_Status_Y'],
      dtype='object')

In [13]:
## Column type check
df.dtypes

ApplicantIncome              int64
CoapplicantIncome          float64
LoanAmount                 float64
Loan_Amount_Term           float64
Credit_History             float64
Loan_Status_Num              int64
Gender_Female                uint8
Gender_Male                  uint8
Married_No                   uint8
Married_Yes                  uint8
Dependents_0                 uint8
Dependents_1                 uint8
Dependents_2                 uint8
Dependents_3+                uint8
Education_Graduate           uint8
Education_Not Graduate       uint8
Self_Employed_No             uint8
Self_Employed_Yes            uint8
Property_Area_Rural          uint8
Property_Area_Semiurban      uint8
Property_Area_Urban          uint8
Loan_Status_N                uint8
Loan_Status_Y                uint8
dtype: object

In [15]:
# Split the data into a training and test set, where the “Loan_Status” column is the target.
x = df.drop(['Loan_Status_Num', 'Loan_Status_N', 'Loan_Status_Y'], axis=1) # All other variables
y = df['Loan_Status_Num'] # Target variable

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

In [16]:
## Data split check
print('Training data shape:')
print(x_train.shape)
print(y_train.shape)

print('Testing data shape:')
print(x_test.shape)
print(y_test.shape)

Training data shape:
(384, 20)
(384,)
Testing data shape:
(96, 20)
(96,)


In [21]:
# Create a pipeline with a min-max scaler and a KNN classifier (see section 15.3 in the Machine Learning with Python Cookbook).
minmax_scale = preprocessing.MinMaxScaler(feature_range=(0,1)) #MinMax
knn = KNeighborsClassifier(n_neighbors = 5, n_jobs=-1) #KNN
pipe = Pipeline([('scaler', minmax_scale), ('knn', knn)]) #Pipeline for MinMax and KNN

In [23]:
# Fit a default KNN classifier to the data with this pipeline. Report the model accuracy on the test set. Note: Fitting a pipeline model works just like fitting a regular model.
pipe.fit(x_train, y_train)

Pipeline(steps=[('scaler', MinMaxScaler()),
                ('knn', KNeighborsClassifier(n_jobs=-1))])

In [26]:
## Applying prediction
y_pred = pipe.predict(x_test)
y_pred

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 0, 1, 1], dtype=int64)

In [29]:
## Score check
print(f'Score Check: {pipe.score(x_test, y_test)}')

Score Check: 0.71875


In [30]:
# Create a search space for your KNN classifier where your “n_neighbors” parameter varies from 1 to 10. (see section 15.3 in the Machine Learning with Python Cookbook).
search_space = [{'knn__n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}]

In [31]:
# Fit a grid search with your pipeline, search space, and 5-fold cross-validation to find the best value for the “n_neighbors” parameter.
ss = StandardScaler() # Standardizer
x_ss = ss.fit_transform(x) # Features
classifier = GridSearchCV(pipe, search_space, cv=5, verbose=0).fit(x_ss, y) # Grid Search
classifier.best_estimator_.get_params()['knn__n_neighbors'] # Best K value

3

In [33]:
# Find the accuracy of the grid search best model on the test set. Note: It is possible that this will not be an improvement over the default model, but likely it will be.
print(f'KNN Hyperparameters: {classifier.best_params_}')
print(f'KNN Score Check: {classifier.best_score_}')

KNN Hyperparameters: {'knn__n_neighbors': 3}
KNN Score Check: 0.7229166666666667


In [34]:
# Now, repeat steps 6 and 7 with the same pipeline, but expand your search space to include logistic regression and random forest models with the 
# hyperparameter values in section 12.3 of the Machine Learning with Python Cookbook.
ss1 = StandardScaler() # Standardizer
classifier1 = KNeighborsClassifier(n_neighbors=5, n_jobs=-1) # KNeighbor Classifier
pipe1 = Pipeline([('standardizer', ss1), ('classifier', classifier1)]) # pipe creation

In [36]:
## Setting up variables for search space
knn1 = KNeighborsClassifier()
cknn1 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
lr = LogisticRegression()
penalty = ['l1', 'l2']
ls = np.logspace(0, 4, 10)
solver = ['liblinear']
forest = RandomForestClassifier()
estimator = [10, 100, 1000]
max_f = [1, 2, 3]

In [37]:
## Applying Search Space
search_space1 = [{'classifier' : [knn1],
                 'classifier__n_neighbors' : cknn1},
                {'classifier' : [lr],
                'classifier__penalty' : penalty,
                'classifier__C' : ls,
                'classifier__solver' : solver},
                {'classifier' : [forest],
                'classifier__n_estimators' : estimator,
                'classifier__max_features' : max_f}]

In [38]:
## Applying Grid Searhc
gridsearch = GridSearchCV(pipe1, search_space1, cv=5, verbose=0)

In [39]:
## Applying fit to grid search
grid_model = gridsearch.fit(x_train, y_train)

In [40]:
# What are the best model and hyperparameters found in the grid search? Find the accuracy of this model on the test set.
grid_model.best_estimator_.get_params() # Best K value

{'memory': None,
 'steps': [('standardizer', StandardScaler()),
  ('classifier', LogisticRegression(penalty='l1', solver='liblinear'))],
 'verbose': False,
 'standardizer': StandardScaler(),
 'classifier': LogisticRegression(penalty='l1', solver='liblinear'),
 'standardizer__copy': True,
 'standardizer__with_mean': True,
 'standardizer__with_std': True,
 'classifier__C': 1.0,
 'classifier__class_weight': None,
 'classifier__dual': False,
 'classifier__fit_intercept': True,
 'classifier__intercept_scaling': 1,
 'classifier__l1_ratio': None,
 'classifier__max_iter': 100,
 'classifier__multi_class': 'auto',
 'classifier__n_jobs': None,
 'classifier__penalty': 'l1',
 'classifier__random_state': None,
 'classifier__solver': 'liblinear',
 'classifier__tol': 0.0001,
 'classifier__verbose': 0,
 'classifier__warm_start': False}

In [41]:
## Accuracy of grid model calculation
grid_model = gridsearch.fit(x_test, y_test)
print(f'Grid Model Accuracy: {grid_model.best_score_}')

Grid Model Accuracy: 0.8342105263157895


It seems that the more the pipeline is expanded the more accurate the results are. As we can see the grid_model got a score of 83.4% while the initial model scored a 71.8% accuracy.